test/Microsoft.AspNet.Mvc.TagHelpers.Test/ScriptTagHelperTest.cs

```diff
@@ -0,0 +1,204 @@
+        }
+
+        [Fact]
+        public async Task DoesNotRunWhenAllRequiredAttributesAreMissing()
+        {
+            // Arrange
+            var context = MakeTagHelperContext();
+            var output = MakeTagHelperOutput("script");
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.NotNull(output.TagName);
+            Assert.False(output.ContentSet);
+
+            Assert.Equal(1, logger.Logged.Count);
+
+            Assert.Equal(LogLevel.Verbose, logger.Logged[0].Item1);
+            Assert.IsType<string>(logger.Logged[0].Item2);
+            Assert.StartsWith("Skipping processing for ScriptTagHelper", (string)logger.Logged[0].Item2);
+        }
+
+        [Fact]
+        public async Task PreservesOrderOfSourceAttributesWhenRun()
+        {
+            // Arrange
+            var context = MakeTagHelperContext(
+                attributes: new Dictionary<string, object>
+                {
+                    { "data-extra", "something"},
+                    { "src", "/blank.js"},
+                    { "asp-fallback-src", "http://www.example.com/blank.js" },
+                    { "asp-fallback-test", "isavailable()" },
+                });
+
+            var output = MakeTagHelperOutput("link",
+                attributes: new Dictionary<string, string>
+                {
+                    { "data-extra", "something"},
+                    { "src", "/blank.js"},
+                });
+
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+                FallbackSrc = "~/blank.js",
+                FallbackTestMethod = "http://www.example.com/blank.js",
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.StartsWith("<script data-extra='something' src='/blank.js'", output.Content);
+            Assert.Equal(0, logger.Logged.Count);
+        }
+
+        private TagHelperContext MakeTagHelperContext(
+            IDictionary<string, object> attributes = null,
+            string content = null)
+        {
+            attributes = attributes ?? new Dictionary<string, object>();
+
+            return new TagHelperContext(attributes, Guid.NewGuid().ToString("N"), () => Task.FromResult(content));
+        }
+
+        private TagHelperOutput MakeTagHelperOutput(string tagName, IDictionary<string, string> attributes = null)
+        {
+            attributes = attributes ?? new Dictionary<string, string>();
+
+            return new TagHelperOutput(tagName, attributes);
+        }
+
+        private class Logger : ILogger<ScriptTagHelper>
```

Comment:
because there is no need to create moq for that. It's ugly and unreadable when it comes to setting up method calls.

I'm not just checking `Logged.Count` it is specifically checking the logged objects